### PR TITLE
Split display

### LINKED ISSUE #5: Optimize DisplayGrid
The current `DisplayGrid` uses a lot of `cout`s. It is faster to create a string and then do one single cout.

### LINKED ISSUE #13: should we let user use display as a string variabe ?
Des la conception on est parti du principe que l'utilisateur dois formater les différents éléments a afficher sous la forme de string.

Pour l'instant il manque une fonction du style appendContent() qui permettrais d'ajouter du contenu en plus et non de le remplacer uniquement.

Dans l'optique d'améliorer l'utilisation de la librairie je trouve qu'il serait intéressant de pouvoir modifier le contenu directement à l'aide des opérateurs ´=´ et ´+=´ qui ferait respectivement un setContent et un appendContent.

ça serait surtout utile dans le cadre ou il y aurait des constructions complexe par exemple si quelqu'un aimerais refaire ça propre grille pour éviter l'utilisation de variable temporaire.

Exemple : 
```c++
// Before
Display display();
string tmp;

vector<vector<string>> grid(6, vector<string>(6, "0"));
  for (auto &row : grid) {
    for (auto &item : row) {
       tpm += item;
    }
    tpm += "\n";
  }
  display.show(tmp);

// After

Display display();

vector<vector<string>> grid(6, vector<string>(6, "0"));
  for (auto &row : grid) {
    for (auto &item : row) {
       display += item;
    }
    display += "\n";
  }
  display.show(); // or cout << display

```

### LINKED ISSUE #20: Split `Display` class
Currently the `Display` class does (at least) two things:
- handle the display
- handle the string to display

One could split these options, then things like #13 would make sense:
- `Display` handles displaying things, cursor movement
- `DisplayString` handles creating a string with colors, for example:
```c++
DisplayString s("Hello world\n");
s.changeColor(BLUE);
s += "How are you?\n";
s.addColorText(GREEN, "I'm fine\n");
// This would then be useful in the `Display.displayGrid` method to evaluate the width of the boxes
assert("Hello world\nHow are you?\nI'm fine\n".length() == s.length());
// Count the "\n" + 1
assert(4, s.lines());
Display d;
d.show(s);
```

### LINKED ISSUE #21: Lowercase `Display::DisplayGrid`
I have no idea why this method is uppercase. It should be lowercase, like the others.

### LINKED ISSUE #27: Refactoring libdio
My proposition is to do the following:

```
namespace Display {
  enum class Color {
    RED = xx,
    ...
  }

  // DString represents a colored string. Upon creation it is initialised with
  // a white text color.
  class DString {
    DString(Color c = Color::WHITE);
    void setColor(Color color);

    DString& operator+=(const DString &ds);
    friend DString operator+(DString ds, const DString &ds);

    template<typedef T>
    DString& operator+=(const T);
    template<typedef T>
    DString operator+(DString ds, const T);

    // Prints the string and clears DString.
    void print();

    void saveCursorPosition();
    void goBackToCursorPosition();
    void clear();

    // Returns the maximum line-length of this string.
    size_t max_width();
    // Returns the number of lines in this string
    size_t count_lines();
  }

  DString displayGrid(const vector<vector<DString>> &grid, bool grid);
  template<typedef T>
  DString displayGrid(const vector<vector<T>> &grid, bool grid);
  template<typedef T>
  DString displayGrid(const vector<vector<T>> &grid, function<DString(T)> convert, bool grid);
}
```

This would close most of the issues around refactoring. What do you think, @Stellucidam @SuperQuentin @A2va ?

### PR DESCRIPTION
**Pull request by ineiti**
- Separated Display into a namespace with a DString class and standalone
- methods to display the grid
- First steps in templating things
- inherit from std::string

Closes #5 
Closes #13 
Closes #20
Closes #21
Closes #27 #25

### CODE REVIEW

File: example/main.cpp
Comment (ineiti) [inline example/main.cpp]: Ceci est très moche - mais il paraît que c'est obligatoire si on veut faire passer une `closure` en tant que `function` dans un `template`...

**Review by Stellucidam — DISMISSED** [review]

File: src/display.cpp
Comment (A2va) [inline src/display.cpp:202]: `displayGrid` is not a template function

Comment (ineiti) [reply src/display.cpp:202]: But I define it here as a template:

https://github.com/HEIGVD-PRG1-F-2022/libdio/pull/29/files#diff-c0ed3f97e01d9b4153f8d89e9c42fc38d233cb82d95ffb178db85576d526bef4R62

File: example/main.cpp
Comment (A2va) [inline example/main.cpp:25]: > If you're already into changing the example/main.cpp, which is copied to the README, you could also remove the sleeps - I don't really think they're useful. And when I'm testing, it's just inconvenient to have to wait for 5 seconds :)

File: src/display.cpp
Comment (Forestierr) [inline src/display.cpp:6]: Hy, I think it is used for system function calls (UNIX). 
You can include ```#include <cstdlib>``` to be sure.

File: include/libdio/display.h
Comment (ineiti) [inline include/libdio/display.h:60]: The `operator<<` is only defined for a `ostream` in `string`:

https://en.cppreference.com/w/cpp/string/basic_string/operator_ltltgtgt

File: example/main.cpp
Comment (ineiti) [reply example/main.cpp:25]: I removed all the other sleeps and only kept this to show the `saveCursorPosition` and `restoreCursorPosition`. But I can also remove the demo of these two methods - it doesn't work on MacOS anyway.

Does it work on Windows? It's supposed to overwrite the string `Hello World`.

**Comment by A2va** [comment]
Inherit `DString` from `std::string` is for now not that much usable because if we insert a character in the string it can be in the middle of an escape sequence. But we can perhaps think about it for a round of refactoring

File: src/display.cpp
Comment (ineiti) [reply src/display.cpp:6]: Added.

File: example/main.cpp
Comment (ineiti) [reply example/main.cpp]: Amélioré dans la nouvelle version.

File: example/main.cpp
Comment (ineiti) [reply example/main.cpp:25]: Found a way to make it work on Linux. But not on MacOS CLion...

**Comment by ineiti** [comment]
> Inherit `DString` from `std::string` is for now not that much usable because if we insert a character in the string it can be in the middle of an escape sequence. But we can perhaps think about it for a round of refactoring

True - also counting characters is off if you don't take into account the ANSI sequences... So I wonder whether it's worth at all to inherit from `std::string`. I honestly don't see the use.

File: src/display.cpp
Comment (A2va) [reply src/display.cpp:202]: I am not an expert on templates, but I tought all the template implementation must be in a tpp file.

File: include/libdio/display.h
Comment (A2va) [reply include/libdio/display.h:60]: It's ok, I got confused

**Comment by A2va** [comment]
Maybe you're right, for now we don't need this inheritance.

**Comment by ineiti** [comment]
Let's keep it using `String` for the moment - if one of you approves the PR, I'll merge it.

@A2va @Stellucidam

**Review by A2va — DISMISSED** [review]

**Comment by A2va** [comment]
All that remains is to resolve the conflict

**Comment by ineiti** [comment]
My bad, @A2va - now you need to approve again after the merge...

**Review by A2va — APPROVED** [review]

**Pull request merged**
